io-classes/src/Control/Monad/Class/MonadUnique.hs

```diff
@@ -0,0 +1,33 @@
+{-# LANGUAGE TypeFamilies, AllowAmbiguousTypes #-}
+
+-- | A generalisation of the
+-- <https://hackage.haskell.org/package/base/docs/Data-Unique.html Data.Unique>
+-- API to both 'IO' and <https://hackage.haskell.org/package/io-sim IOSim>.
+--
+module Control.Monad.Class.MonadUnique (
+  MonadUnique(..)
+) where
+
+-- base
+import Data.Kind (Type)
+import Data.Unique qualified as IO
+
+-- transformers
+import Control.Monad.Reader (ReaderT(..), lift)
+
+
+class (Monad m, Eq (Unique m), Ord (Unique m)) => MonadUnique m where
+  type Unique m :: Type
+  newUnique  :: m (Unique m)
+  hashUnique :: Unique m -> Int
```

Comment:
The `hashUnique` cannot be used in this form, because GHC has no way to find out `m` from `Unique m` (turning on `AllowAmbiguousTypes` was not a good idea, as in this case it indicates a real problem).

There are two ways of fixing this, either adding `Proxy` arguments, but that changes API, or using `TypeFamilyDependencies`.  In `io-classes` we prefer the latter (to stay as close to `base` API as possible), although it requires some newtype wrappers for instances (e.g. monad transformers).
```suggestion
class (Monad m, Eq (Unique m), Ord (Unique m)) => MonadUnique m where
  type Unique m = (unique :: Type) | unique -> m
  newUnique  :: m (Unique m)
  hashUnique :: Unique m -> Int
```

Reply:
I had originally intended to mention/ask about this, but it slipped my mind. When choosing between an ambiguous type and an injective type family, I begrudgingly settled for the former to allow the `ReaderT` instance through, but I guess necessary type applications are as big of an API change as added proxies, so it's no good indeed. I don't know why the `newtype` injectivity fix didn't occur to me ... Done.

Reply:
Thanks